driver-core/src/main/com/mongodb/internal/connection/AuthorizationHeader.java

```diff
@@ -0,0 +1,220 @@
+        requestHeaders.put("X-MongoDB-GS2-CB-Flag", "n");
+
+        final String headers = getCanonicalHeaders(requestHeaders);
+        final String signedHeaders = getSignedHeaders(requestHeaders);
+        String hash = hash(body);
+
+        final List<String> request = Arrays.asList(method, uri, "", headers, signedHeaders, hash);
+        return request.stream().collect(Collectors.joining("\n"));
+    }
+
+    public String getSignedHeaders(final Map<String, String> requestHeaders) {
+        return requestHeaders.entrySet().stream()
+                .map(kvp -> kvp.getKey())
+                .map(String::toLowerCase)
+                .sorted()
+                .collect(Collectors.joining(";"));
+    }
+
+    public String getCanonicalHeaders(final Map<String, String> requestHeaders) {
+        return requestHeaders.entrySet().stream()
+                .map(kvp -> String.format("%s:%s", kvp.getKey().toLowerCase(), kvp.getValue()))
+                .sorted()
+                .collect(Collectors.joining("\n", "", "\n"));
+    }
+
+    public String getRegion(final String host) throws SaslException {
+        if (host.matches("(\\w)+") || host.equals("sts.amazonaws.com")) {
+            return "us-east-1"; /* default */
+        }
+        if (host.matches("(\\w)+.(\\w+)")) {
+            return host.substring(host.indexOf('.') + 1);
+        }
+        throw new SaslException("Invalid server host");
+    }
+
+    public String getSignature() {
+        return this.signature;
+    }
+
+    public String getTimestamp() {
+        return this.timestamp;
+    }
+
+    private String getDate() {
+        return getTimestamp().substring(0, "YYYYMMDD".length());
+    }
+
+    public String getCanonicalQuery(final Map<String, String> parameters) {
+        String body = parameters.entrySet().stream()
+                .map(map -> {
+                    if (map.getValue() == null) {
+                        return String.format("%s", map.getKey());
+                    }
+                    return String.format("%s=%s", map.getKey(), map.getValue());
+                })
+                .sorted()
+                .collect(Collectors.joining("&"));
+
+        return String.format("Action=%s", body);
+    }
+
+    public String hash(final String str) throws SaslException {
+        return hexEncode(sha256(str)).toLowerCase();
+    }
+
+    private String hexEncode(final byte[] bytes) {
+        StringBuilder sb = new StringBuilder();
+        for (byte b : bytes) {
+            sb.append(String.format("%02x", b));
+        }
+        return sb.toString();
+    }
+
+
+    private byte[] decodeUTF8(final String str) throws SaslException {
+        try {
+            return str.getBytes("UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            throw new SaslException("UTF-8 is not a supported encoding.", e);
+        }
+    }
+
+    private byte[] hmac(final byte[] secret, final byte[] message) throws SaslException {
```

Comment:
you can make this and the hash function `static`

Reply:
Done.